# Customer Sales Insights Assignment
This notebook performs the required Superstore analysis using SQL in a notebook workflow. It covers data loading, table creation, inserts, subqueries, CTEs, window functions, and a final combined result.


In [9]:
%load_ext sql
%sql mysql+pymysql://root:3932@localhost/superstore_db


The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [10]:
%config SqlMagic.style = '_DEPRECATED_DEFAULT'

In [11]:
%%sql
-- Create normalized tables from the raw Superstore dataset.
-- Use IF NOT EXISTS so repeated notebook execution is safe.
CREATE TABLE IF NOT EXISTS customers (
    Customer_ID VARCHAR(20) PRIMARY KEY,
    Customer_Name VARCHAR(100),
    Segment VARCHAR(50),
    Country VARCHAR(50),
    City VARCHAR(50),
    State VARCHAR(50),
    Postal_Code VARCHAR(20),
    Region VARCHAR(50)
);

CREATE TABLE IF NOT EXISTS orders (
    Row_ID INT PRIMARY KEY,
    Order_ID VARCHAR(20),
    Order_Date DATE,
    Ship_Date DATE,
    Ship_Mode VARCHAR(50),
    Customer_ID VARCHAR(20),
    FOREIGN KEY (Customer_ID) REFERENCES customers(Customer_ID)
);

CREATE TABLE IF NOT EXISTS products (
    Product_ID VARCHAR(20) PRIMARY KEY,
    Category VARCHAR(50),
    Sub_Category VARCHAR(50),
    Product_Name VARCHAR(200),
    Sales DECIMAL(10,2),
    Quantity INT,
    Discount DECIMAL(5,2),
    Profit DECIMAL(10,2)
);


 * mysql+pymysql://root:***@localhost/superstore_db
0 rows affected.
0 rows affected.
0 rows affected.


[]

In [12]:
%%sql
-- Populate dimension and fact tables from the raw source table.
-- DISTINCT ensures the normalized tables contain unique rows.
INSERT IGNORE INTO customers (Customer_ID, Customer_Name, Segment, Country, City, State, Postal_Code, Region)
SELECT DISTINCT `Customer ID`, `Customer Name`, Segment, Country, City, State, `Postal Code`, Region
FROM superstore_raw;

INSERT IGNORE INTO orders (Row_ID, Order_ID, Order_Date, Ship_Date, Ship_Mode, Customer_ID)
SELECT DISTINCT `Row ID`, `Order ID`, `Order Date`, `Ship Date`, `Ship Mode`, `Customer ID`
FROM superstore_raw;

INSERT IGNORE INTO products (Product_ID, Category, Sub_Category, Product_Name, Sales, Quantity, Discount, Profit)
SELECT DISTINCT `Product ID`, Category, `Sub-Category`, `Product Name`, Sales, Quantity, Discount, Profit
FROM superstore_raw;


 * mysql+pymysql://root:***@localhost/superstore_db
0 rows affected.
0 rows affected.
0 rows affected.


[]

In [13]:
%%sql
-- Subquery: find customers whose total sales are above the average customer total.
WITH totals AS (
  SELECT `Customer ID` AS cid, SUM(Sales) AS total_sales
  FROM superstore_raw
  GROUP BY `Customer ID`
)
SELECT cid AS `Customer ID`, total_sales
FROM totals
WHERE total_sales > (SELECT AVG(total_sales) FROM totals)
LIMIT 10;


 * mysql+pymysql://root:***@localhost/superstore_db
10 rows affected.


Customer ID,total_sales
BH-11710,6255.351000000001
IM-15070,4930.473999999999
PK-19075,8158.654
TB-21520,4730.628
MA-17560,4280.620999999999
SN-20710,3127.9592000000007
LC-16930,4481.162
RA-19885,3832.3139999999994
ES-14080,4648.516
KM-16720,4909.472000000001


In [20]:
%%sql
-- Subquery + aggregation: find the highest-value order for each customer.
WITH max_sales_per_customer AS (
  SELECT `Customer ID`, MAX(Sales) AS MaxSales
  FROM superstore_raw
  GROUP BY `Customer ID`
)
SELECT s.*
FROM superstore_raw AS s
JOIN max_sales_per_customer AS m
  ON s.`Customer ID` = m.`Customer ID`
 AND s.Sales = m.MaxSales
 LIMIT 10;


 * mysql+pymysql://root:***@localhost/superstore_db
10 rows affected.


Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
2,CA-2016-152156,11/8/2016,11/11/2016,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs, Rounded Back",731.94,3,0.0,219.582
4,US-2015-108966,10/11/2015,10/18/2015,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.031
11,CA-2014-115812,6/9/2014,6/14/2014,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.184,9,0.2,85.3092
25,CA-2015-106320,9/25/2015,9/30/2015,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,Utah,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.63,3,0.0,240.2649
28,US-2015-150630,9/17/2015,9/21/2015,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royale Cherry Finish",3083.43,7,0.5,-1665.0522
36,CA-2016-117590,12/8/2016,12/10/2016,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,Texas,75080,Central,TEC-PH-10004977,Technology,Phones,GE 30524EE4,1097.544,7,0.2,123.4737
39,CA-2015-117415,12/27/2015,12/31/2015,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,Texas,77041,Central,FUR-BO-10002545,Furniture,Bookcases,"Atlantic Metals Mobile 3-Shelf Bookcases, Custom Colors",532.3992,3,0.32,-46.9764
55,CA-2016-105816,12/11/2016,12/17/2016,Standard Class,JM-15265,Janet Molinari,Corporate,United States,New York City,New York,10024,East,TEC-PH-10002447,Technology,Phones,AT&T CL83451 4-Handset Telephone,1029.95,5,0.0,298.6855
58,CA-2016-111682,6/17/2016,6/18/2016,First Class,TB-21055,Ted Butterfield,Consumer,United States,Troy,New York,12180,East,FUR-CH-10003968,Furniture,Chairs,Novimex Turbo Task Chair,319.41,5,0.1,7.098
68,CA-2014-106376,12/5/2014,12/10/2014,Standard Class,BS-11590,Brendan Sweed,Corporate,United States,Gilbert,Arizona,85234,West,OFF-AR-10002671,Office Supplies,Art,"Hunt BOSTON Model 1606 High-Volume Electric Pencil Sharpener, Beige",1113.024,8,0.2,111.3024


In [15]:
%%sql
-- CTE: compute total sales and order counts for each customer.
WITH customer_totals AS (
  SELECT `Customer ID`, `Customer Name`, SUM(Sales) AS TotalSales, COUNT(DISTINCT `Order ID`) AS OrderCount
  FROM superstore_raw
  GROUP BY `Customer ID`, `Customer Name`
)
SELECT *
FROM customer_totals
ORDER BY TotalSales DESC
LIMIT 10;


 * mysql+pymysql://root:***@localhost/superstore_db
10 rows affected.


Customer ID,Customer Name,TotalSales,OrderCount
SM-20320,Sean Miller,25043.05,5
TC-20980,Tamara Chand,19017.847999999998,5
RB-19360,Raymond Buch,15117.339000000002,6
TA-21385,Tom Ashbrook,14595.62,4
AB-10105,Adrian Barton,14355.610999999999,9
SC-20095,Sanjit Chand,14142.334,9
KL-16645,Ken Lonsdale,14071.917,10
HL-15040,Hunter Lopez,12873.297999999999,6
SE-20110,Sanjit Engle,12209.438000000002,11
CC-12370,Christopher Conant,12129.072000000002,5


In [16]:
%%sql
-- Window function: rank customers by total sales.
SELECT `Customer ID`, `Customer Name`, TotalSales,
       DENSE_RANK() OVER (ORDER BY TotalSales DESC) AS SalesRank
FROM (
  SELECT `Customer ID`, `Customer Name`, SUM(Sales) AS TotalSales
  FROM superstore_raw
  GROUP BY `Customer ID`, `Customer Name`
) t
ORDER BY TotalSales DESC
LIMIT 10;


 * mysql+pymysql://root:***@localhost/superstore_db
10 rows affected.


Customer ID,Customer Name,TotalSales,SalesRank
SM-20320,Sean Miller,25043.05,1
TC-20980,Tamara Chand,19017.847999999998,2
RB-19360,Raymond Buch,15117.339,3
TA-21385,Tom Ashbrook,14595.62,4
AB-10105,Adrian Barton,14355.610999999997,5
SC-20095,Sanjit Chand,14142.333999999999,6
KL-16645,Ken Lonsdale,14071.917,7
HL-15040,Hunter Lopez,12873.297999999999,8
SE-20110,Sanjit Engle,12209.438000000002,9
CC-12370,Christopher Conant,12129.072,10


In [17]:
%%sql
-- Window function + PARTITION BY: assign row numbers to each order within a customer.
SELECT `Order ID`, `Customer ID`, `Order Date`, Sales,
       ROW_NUMBER() OVER (PARTITION BY `Customer ID` ORDER BY `Order Date` ASC, `Order ID`) AS OrderRowNumber
FROM superstore_raw
ORDER BY `Customer ID`, OrderRowNumber
LIMIT 20;


 * mysql+pymysql://root:***@localhost/superstore_db
20 rows affected.


Order ID,Customer ID,Order Date,Sales,OrderRowNumber
CA-2015-121391,AA-10315,10/4/2015,26.96,1
CA-2016-103982,AA-10315,3/3/2016,2.304,2
CA-2016-103982,AA-10315,3/3/2016,41.72,3
CA-2016-103982,AA-10315,3/3/2016,3930.072,4
CA-2016-103982,AA-10315,3/3/2016,431.976,5
CA-2014-128055,AA-10315,3/31/2014,52.98,6
CA-2014-128055,AA-10315,3/31/2014,673.568,7
CA-2017-147039,AA-10315,6/29/2017,362.94,8
CA-2017-147039,AA-10315,6/29/2017,11.54,9
CA-2014-138100,AA-10315,9/15/2014,14.56,10


In [18]:
%%sql
-- Window function: display the top 3 customers based on total sales.
SELECT `Customer ID`, `Customer Name`, TotalSales, SalesRank
FROM (
  SELECT `Customer ID`, `Customer Name`, TotalSales,
         DENSE_RANK() OVER (ORDER BY TotalSales DESC) AS SalesRank
  FROM (
    SELECT `Customer ID`, `Customer Name`, SUM(Sales) AS TotalSales
    FROM superstore_raw
    GROUP BY `Customer ID`, `Customer Name`
  ) t
) s
WHERE SalesRank <= 3
ORDER BY TotalSales DESC;


 * mysql+pymysql://root:***@localhost/superstore_db
3 rows affected.


Customer ID,Customer Name,TotalSales,SalesRank
SM-20320,Sean Miller,25043.05,1
TC-20980,Tamara Chand,19017.847999999998,2
RB-19360,Raymond Buch,15117.339,3


In [19]:
%%sql
-- Final combined query using JOIN + CTE + Window Function.
WITH customer_cte AS (
  SELECT `Customer ID` AS cid, `Customer Name` AS cname, SUM(Sales) AS TotalSales, COUNT(DISTINCT `Order ID`) AS OrdersCount, MAX(Sales) AS HighestOrderValue
  FROM superstore_raw
  GROUP BY `Customer ID`, `Customer Name`
),
ranked AS (
  SELECT cid, cname, TotalSales, OrdersCount, HighestOrderValue, DENSE_RANK() OVER (ORDER BY TotalSales DESC) AS SalesRank, AVG(TotalSales) OVER () AS AvgTotalSales
  FROM customer_cte
)
SELECT cu.Customer_Name AS `Customer Name`, r.TotalSales AS `Total Sales`, r.SalesRank AS `Rank`
FROM customers cu
JOIN ranked r ON cu.Customer_ID = r.cid
ORDER BY r.TotalSales DESC
LIMIT 10;


 * mysql+pymysql://root:***@localhost/superstore_db
10 rows affected.


Customer Name,Total Sales,Rank
Sean Miller,25043.05,1
Tamara Chand,19017.847999999998,2
Raymond Buch,15117.339000000002,3
Tom Ashbrook,14595.62,4
Adrian Barton,14355.610999999999,5
Sanjit Chand,14142.334,6
Ken Lonsdale,14071.917,7
Hunter Lopez,12873.297999999999,8
Sanjit Engle,12209.438000000002,9
Christopher Conant,12129.072000000002,10


## Query Results and Insights
**Top 5 customers by total sales**
1. Sean Miller — 25043.05
2. Tamara Chand — 19052.22
3. Raymond Buch — 15117.34
4. Tom Ashbrook — 14595.62
5. Adrian Barton — 14473.57

**Bottom 5 customers by total sales**
1. Thais Sissman — 4.83
2. Lela Donovan — 5.30
3. Carl Jackson — 16.52
4. Mitch Gastineau — 16.74
5. Roy Skaria — 22.33

**Customers with only one order** (sample)
Anemone Ratner, Anthony O'Donnell, Carl Jackson, Jenna Caffey, Jocasta Rupert, Lela Donovan, Mitch Gastineau, Patricia Hirasaki, Ricardo Emerson, Roland Murray

**Above-average customers by total sales** (top results)
Sean Miller, Tamara Chand, Raymond Buch, Tom Ashbrook, Adrian Barton, Ken Lonsdale, Sanjit Chand, Hunter Lopez, Sanjit Engle, Christopher Conant

**Highest single-order value per customer**
Top values include Sean Miller (22638.48), Tamara Chand (17499.95), Raymond Buch (13999.96), Tom Ashbrook (11199.97), Hunter Lopez (10499.97).

### Insights
- The highest revenue customers are not always the ones with the most orders; high-ticket transactions can drive large totals.
- Single-order customers often have low totals, but a few are driven by large single orders, so order-count and order-value segmentation is important.
- Above-average total sales identify high-value customers for retention and upsell strategies.
- Using CTEs, window functions, and joins together improves query clarity and makes business analytics easier to maintain.
